awx/main/notifications/webhook_backend.py

```diff
@@ -61,23 +63,60 @@ def format_body(self, body):
             raise ValueError("HTTP method must be either 'POST' or 'PUT'.")
         chosen_method = getattr(requests, self.http_method.lower(), None)
+        resp = None
+        '''
+        Sometimes we hit redirect URLs. We must account for this. We still extract the redirect URL from the response headers and try again. Max retires == 5
+        '''
         for m in messages:
+
             auth = None
             if self.username or self.password:
                 auth = (self.username, self.password)
-            r = chosen_method(
-                "{}".format(m.recipients()[0]),
-                auth=auth,
-                data=json.dumps(m.body, ensure_ascii=False).encode('utf-8'),
-                headers=dict(list(get_awx_http_client_headers().items()) + list((self.headers or {}).items())),
-                verify=(not self.disable_ssl_verification),
-            )
-            if r.status_code >= 400:
-                logger.error(smart_str(_("Error sending notification webhook: {}").format(r.status_code)))
-                if not self.fail_silently:
-                    raise Exception(smart_str(_("Error sending notification webhook: {}").format(r.status_code)))
-            sent_messages += 1
+
+            # convert emtpy string to empty dict for POST/PUT
+            if not m.body:
+                m.body = dict()
+
+            url = "{}".format(m.recipients()[0])
+            data = json.dumps(m.body, ensure_ascii=False).encode('utf-8')
+
+            headers = dict(list(get_awx_http_client_headers().items()) + list((self.headers or {}).items()))
+
+            for retries in range(self.MAX_RETIRES):
+
+                resp = chosen_method(
+                    url=url,
+                    auth=auth,
+                    data=data,
+                    headers=headers,
+                    verify=(not self.disable_ssl_verification),
+                    allow_redirects=False,  # override default behaviour for redirects
+                )
+
+                # we've hit a redirect. extract the redirect URL out of the first response header and try again
+                if resp.status_code in [301, 307]:
+                    # take the first redirect URL in the response header and try that
+                    url = resp.headers.get("Location", None)
+                    if url is None:
+                        raise Exception(smart_str(_("Recieved redirect to a blank URL. Response headers={}").format(resp_inner.history[0].headers)))
+                    continue
+
+                if resp.status_code in range(200, 299):
+                    sent_messages += 1
+                    break
+
+                if resp.status_code >= 400:
+                    logger.error(smart_str(_("Error sending notification webhook: {}").format(resp.status_code)))
+                    if not self.fail_silently:
+                        raise Exception(smart_str(_("Error sending notification webhook: {}").format(resp.status_code)))
```

Comment:
and here

Reply:
I suppose it depends on whether some caller upstream is catching the exception and doing some logging or whatever with it.  :man_shrugging: